File: fabric/src/main/java/juuxel/adorn/util/FluidStorageReference.java
package juuxel.adorn.util;

import juuxel.adorn.fluid.FluidReference;
import juuxel.adorn.fluid.FluidUnit;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidVariant;
import net.fabricmc.fabric.api.transfer.v1.storage.base.SingleVariantStorage;
import net.minecraft.core.component.DataComponentPatch;
import net.minecraft.world.level.material.Fluid;

/**
 * A {@linkplain FluidReference fluid reference} to a {@link SingleVariantStorage}.
 */
public final class FluidStorageReference extends FluidReference {
    private final SingleVariantStorage<FluidVariant> storage;

    public FluidStorageReference(SingleVariantStorage<FluidVariant> storage) {
        this.storage = storage;
    }

    public FluidVariant getVariant() {
        return storage.variant;
    }

    @Override
    public Fluid getFluid() {
        return storage.variant.getFluid();
    }

    @Override
    public void setFluid(Fluid fluid, long amount, DataComponentPatch components) {
        storage.variant = FluidVariant.of(fluid, components);
        setAmount(amount);
    }

    @Override
    public long getAmount() {
        return storage.amount;
    }

    @Override
    public void setAmount(long amount) {
        storage.amount = amount;
    }

    @Override
    public DataComponentPatch getComponents() {
        return storage.variant.getComponentsPatch();
    }

    @Override
    public FluidUnit getUnit() {
        return FluidUnit.DROPLET;
    }

    /**
     * Converts this fluid reference to a {@link FluidVariant}.
     * This is faster than a manual conversion for a {@code FluidStorageReference}.
     */
    public static FluidVariant toFluidVariant(FluidReference reference) {
        if (reference instanceof FluidStorageReference fsr) {
            return fsr.getVariant();
        } else {
            return FluidVariant.of(reference.getFluid(), reference.getComponents());
        }
    }
}

File: fabric/src/main/kotlin/juuxel/adorn/util/FluidStorageReference.kt
package juuxel.adorn.util

import juuxel.adorn.fluid.FluidReference
import juuxel.adorn.fluid.FluidUnit
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidVariant
import net.fabricmc.fabric.api.transfer.v1.storage.base.SingleVariantStorage
import net.minecraft.fluid.Fluid
import net.minecraft.nbt.NbtCompound

/**
 * A [fluid reference][FluidReference] to a [`SingleVariantStorage<FluidVariant>`][SingleVariantStorage].
 */
class FluidStorageReference(private val storage: SingleVariantStorage<FluidVariant>) : FluidReference() {
    val variant: FluidVariant get() = storage.variant

    override var fluid: Fluid
        get() = storage.variant.fluid
        set(value) {
            storage.variant = FluidVariant.of(value, storage.variant.nbt)
        }

    override var amount: Long
        get() = storage.amount
        set(value) {
            storage.amount = value
        }

    override var nbt: NbtCompound?
        get() = storage.variant.nbt
        set(value) {
            storage.variant = FluidVariant.of(storage.variant.fluid, value)
        }

    override val unit = FluidUnit.DROPLET
}
